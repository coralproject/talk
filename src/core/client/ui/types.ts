import React from "react";

// TODO: Extract useful common types into its own package.

/**
 * Overwrite properties of `T`.
 *
 * E.g. Omit<{a: boolean, b: boolean}, "b"> = {a: boolean}
 */
export type Omit<U, K extends keyof U> = Pick<U, Exclude<keyof U, K>>;

/**
 * Overwrite properties of `T`.
 *
 * E.g. Overwrite<{a: boolean}, {a: string}> = {a: string}
 */
export type Overwrite<T, U> = Pick<T, Exclude<keyof T, keyof U>> & U;

/**
 * Returns the PropTypes from a React Component.
 *
 * E.g. type ButtonProps = ReturnPropTypes<Button>;
 */
export type PropTypesOf<T> = T extends React.ComponentType<infer R>
  ? R
  : T extends React.Component<infer S> ? S : {};
