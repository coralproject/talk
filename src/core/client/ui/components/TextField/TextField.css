.root {
  composes: inputText placeholderPseudo from "talk-ui/shared/typography.css";
  position: relative;
  display: block;
  padding: calc(0.5 * var(--spacing-unit));
  box-sizing: border-box;
  border-radius: var(--round-corners);
  height: 36px;
  line-height: 36px;
  width: calc(29 * var(--spacing-unit));

  &:read-only {
    background-color: var(--palette-grey-lightest);
  }
  &:disabled {
    color: var(--palette-text-secondary);
    background-color: var(--palette-grey-lightest);
  }
}

.colorRegular {
  background-color: var(--palette-common-white);
  color: var(--palette-common-black);
  border: 1px solid var(--palette-grey-light);
}

.colorError {
  background-color: var(--palette-common-white);
  border-color: var(--palette-error-main);
  border: 2px solid var(--palette-error-darkest);
}

.fullWidth {
  width: 100%;
}
