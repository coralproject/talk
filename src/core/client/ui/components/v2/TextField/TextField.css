$input-value: var(--v2-colors-mono-500);
$input-border: var(--v2-colors-grey-500);
$input-background: var(--v2-colors-pure-white);
$input-placeholder: var(--v2-colors-mono-100);
$input-regular-focused-border: var(--v2-colors-teal-500);
$input-disabled-background: var(--v2-colors-white-500);
$input-disabled-value: var(--v2-colors-grey-400);
$input-disabled-border: var(--v2-colors-grey-400);
$input-error-border: var(--v2-palette-error-darkest);
$input-dark-focused-border: var(--v2-colors-blue-500);

$input-streamError-border: var(--v2-colors-red-500);

.root {
  display: flex;
  width: calc(29 * var(--mini-unit));
  align-items: center;
  height: 34px;
}

.input {
  font-family: var(--v2-font-family-primary);
  font-weight: var(--v2-font-weight-primary-regular);
  font-size: var(--v2-font-size-3);
  line-height: var(--v2-line-height-reset);
  position: relative;
  display: block;
  padding: var(--v2-spacing-2);
  box-sizing: border-box;
  background-color: $input-background;
  border-radius: var(--round-corners);
  width: 100%;
  align-self: stretch;
  margin-top: 0;
  margin-bottom: 0;
  color: $input-value;

  &:read-only,
  &:disabled {
    background-color: $input-disabled-background;
    border-color: $input-disabled-border;
    color: $input-disabled-value;
  }
  &:focus {
    outline: none;
  }
  &::placeholder {
    color: $input-placeholder;
  }
}

.adornment {
  margin-left: var(--v2-spacing-2);

  font-family: var(--v2-font-family-primary);
  font-weight: var(--v2-font-weight-primary-regular);
  font-size: var(--v2-font-size-3);
  line-height: var(--v2-line-height-reset);
}

.colorRegular {
  border: 1px solid $input-border;
  &:focus {
    border: 1px solid $input-regular-focused-border;
    box-shadow: inset 0px 0px 0px 1px $input-regular-focused-border;
  }
}

.colorDark {
  border: 1px solid $input-border;
  &:focus {
    border: 1px solid $input-dark-focused-border;
    box-shadow: inset 0px 0px 0px 1px $input-dark-focused-border;
  }
}

.colorError {
  border: 2px solid $input-error-border;
}

.colorStreamError {
  border: 2px solid $input-streamError-border;
}

.fullWidth {
  width: 100%;
}

.textAlignCenter {
  text-align: center;
}

.seamlessAdornment {
  border-right: 0;
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;

  &:focus {
    border-right: 0;
  }

  & + .adornment {
    height: 100%;
    margin: 0;
    display: flex;
    align-items: stretch;
  }
}
