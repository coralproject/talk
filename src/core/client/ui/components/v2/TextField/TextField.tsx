import cn from "classnames";
import React, {
  AllHTMLAttributes,
  ChangeEvent,
  EventHandler,
  FunctionComponent,
} from "react";

import { withStyles } from "coral-ui/hocs";

import styles from "./TextField.css";

export interface TextFieldProps {
  id?: string;
  /**
   * The content value of the component.
   */
  defaultValue?: string;
  /**
   * The content value of the component.
   */
  value?: string;
  /**
   * Convenient prop to override the root styling.
   */
  className?: string;
  /**
   * Override or extend the styles applied to the component.
   */
  classes: typeof styles;
  /**
   * Color of the TextField
   */
  color?: "regular" | "error" | "streamError" | "dark";
  /*
   * If set renders a full width button
   */
  fullWidth?: boolean;
  /**
   * Placeholder
   */
  placeholder?: string;
  /**
   * Mark as readonly
   */
  readOnly?: boolean;
  /**
   * Name
   */
  name?: string;
  /**
   * type: Here we only allow text type values
   */
  type?:
    | "email"
    | "number"
    | "password"
    | "search"
    | "tel"
    | "text"
    | "url"
    | string;
  /**
   * onChange
   */
  onChange?: EventHandler<ChangeEvent<HTMLInputElement>>;
  onKeyPress?: React.EventHandler<React.KeyboardEvent>;
  onKeyDown?: React.EventHandler<React.KeyboardEvent>;

  disabled?: boolean;

  autoComplete?: AllHTMLAttributes<HTMLInputElement>["autoComplete"];
  autoCorrect?: AllHTMLAttributes<HTMLInputElement>["autoCorrect"];
  autoCapitalize?: AllHTMLAttributes<HTMLInputElement>["autoCapitalize"];
  spellCheck?: AllHTMLAttributes<HTMLInputElement>["spellCheck"];

  textAlignCenter?: boolean;
  adornment?: React.ReactNode;

  variant?: "regular" | "seamlessAdornment";
}

const TextField: FunctionComponent<TextFieldProps> = (props) => {
  const {
    className,
    classes,
    color,
    fullWidth,
    value,
    placeholder,
    adornment,
    textAlignCenter,
    variant,
    ...rest
  } = props;

  const rootClassName = cn(
    classes.root,
    {
      [classes.fullWidth]: fullWidth,
    },
    className
  );

  const inputClassName = cn(classes.input, {
    [classes.colorRegular]: color === "regular",
    [classes.colorError]: color === "error",
    [classes.colorStreamError]: color === "streamError",
    [classes.colorDark]: color === "dark",
    [classes.textAlignCenter]: textAlignCenter,
    [classes.seamlessAdornment]: variant === "seamlessAdornment",
  });

  return (
    <div className={rootClassName}>
      <input
        className={inputClassName}
        placeholder={placeholder}
        value={value}
        {...rest}
      />
      {adornment && <div className={styles.adornment}>{adornment}</div>}
    </div>
  );
};

TextField.defaultProps = {
  color: "regular",
  placeholder: "",
  type: "text",
} as Partial<TextFieldProps>;

const enhanced = withStyles(styles)(TextField);
export default enhanced;
