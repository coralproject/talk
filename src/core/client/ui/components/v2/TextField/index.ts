export * from "./TextField";
export { default } from "./TextField";
