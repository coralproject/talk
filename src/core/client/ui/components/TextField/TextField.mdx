---
name: TextField
menu: UI Kit
---

import { Playground, PropsTable } from 'docz'
import TextField from './TextField.tsx'
import Flex from '../Flex'

# TextField

## Basic Use
<Playground>
  <Flex itemGutter direction="column">
    <TextField placeholder="This is a placeholder" />
    <TextField defaultValue="This is an input field" />
    <TextField color="error" defaultValue="A TextField with an error" />
    <TextField color="error" defaultValue="A TextField with an error" fullWidth/>
  </Flex>
</Playground>
