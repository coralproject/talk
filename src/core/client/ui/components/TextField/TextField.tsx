import cn from "classnames";
import React, { ChangeEvent, EventHandler } from "react";
import { StatelessComponent } from "react";
import { withStyles } from "talk-ui/hocs";
import * as styles from "./TextField.css";

export interface TextFieldProps {
  id?: string;
  /**
   * The content value of the component.
   */
  defaultValue?: string;
  /**
   * The content value of the component.
   */
  value?: string;
  /**
   * Convenient prop to override the root styling.
   */
  className?: string;
  /**
   * Override or extend the styles applied to the component.
   */
  classes: typeof styles;
  /**
   * Color of the TextField
   */
  color?: "regular" | "error";
  /*
  * If set renders a full width button
  */
  fullWidth?: boolean;
  /**
   * Placeholder
   */
  placeholder?: string;
  /**
   * Mark as readonly
   */
  readOnly?: boolean;
  /**
   * Name
   */
  name?: string;
  /**
   * type: Here we only allow text type values
   */
  type?: "email" | "number" | "password" | "search" | "tel" | "text" | "url";
  /**
   * onChange
   */
  onChange?: EventHandler<ChangeEvent<HTMLInputElement>>;

  disabled?: boolean;
}

const TextField: StatelessComponent<TextFieldProps> = props => {
  const {
    className,
    classes,
    color,
    fullWidth,
    value,
    placeholder,
    ...rest
  } = props;

  const rootClassName = cn(
    classes.root,
    {
      [classes.colorRegular]: color === "regular",
      [classes.colorError]: color === "error",
      [classes.fullWidth]: fullWidth,
    },
    className
  );

  return (
    <input
      className={rootClassName}
      placeholder={placeholder}
      value={value}
      {...rest}
    />
  );
};

TextField.defaultProps = {
  color: "regular",
  placeholder: "",
  type: "text",
};

const enhanced = withStyles(styles)(TextField);
export default enhanced;
