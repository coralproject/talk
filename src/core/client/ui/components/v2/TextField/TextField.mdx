---
name: TextField
menu: UI Kit
---

import { Playground, PropsTable } from "docz";

import TextField from "./TextField";
import Button from "../Button";
import Icon from "../Icon";
import Flex from "../Flex";
import HorizontalGutter from "../HorizontalGutter";

# TextField

## Basic Use

<Playground>
  <HorizontalGutter>
    <Flex>
      <TextField
        placeholder="This is a placeholder"
        variant="seamlessAdornment"
        adornment={
          <Button variant="adornment" color="dark">
            <Icon size="md">search</Icon>
          </Button>
        }
      />
    </Flex>
    <TextField defaultValue="This is an input field" />
    <TextField color="error" defaultValue="A TextField with an error" />
    <TextField
      color="error"
      defaultValue="A TextField with an error"
      fullWidth
    />
    <TextField
      placeholder="This is a placeholder"
      adornment={<span>Seconds</span>}
    />
  </HorizontalGutter>
</Playground>
