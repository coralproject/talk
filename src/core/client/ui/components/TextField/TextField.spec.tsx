import React from "react";
import TestRenderer from "react-test-renderer";

import { PropTypesOf } from "talk-ui/types";

import TextField from "./TextField";

it("renders correctly", () => {
  const props: PropTypesOf<typeof TextField> = {
    className: "custom",
    defaultValue: "Hello World",
  };
  const renderer = TestRenderer.create(<TextField {...props} />);
  expect(renderer.toJSON()).toMatchSnapshot();
});
