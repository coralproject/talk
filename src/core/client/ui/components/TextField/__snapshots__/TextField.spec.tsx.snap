// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<input
  className="TextField-root TextField-colorRegular custom"
  defaultValue="Hello World"
  placeholder=""
/>
`;
